### ISSUE TITLE
Binary encoding support (lets some types have binary persistence)

### ISSUE THREAD

**Issue by abierman** [netmod-wg/yang-next#46]
The YANG to CBOR encoding algorithms use binary types for some YANG data types.
For example, enumerations are sent using the value number and bits are sent as a
number with bits set corresponding to the position number.

There are string types (ietf-inet-types) that CBOR would like to encode as a number.
If possible, this should be done with a YANG statement instead of hard-wired in YANG to CBOR

There are some corner-cases in the "union" data type that do not work

- union of 2 enum types where value numbers overlap

- union of number and enumeration where enum value and number overlap

There could be usage guidelines to prevent overlap.
Perhaps automatic adjustment in the YANG compiler.

**Comment by rgwilton** [comment]
I agree that it would be good to allow some of the types currently encoded as a string to also be encoded numerically.  Having a YANG statement to express this (perhaps initially as an extension) seems like a reasonable approach.

For the union case:
 - The decoder just select the first type for which the value is valid.  So the behaviour is always deterministic at least.
 - CBOR also use a tag for some union types, so it can distinguish between enum and number, but not two enums with the same value.

**Comment by mbj4668** [comment]
I assume this is not just for enums but for strings like ipv4-address and dotted-quad?  It may make sense to add something like this, but the problem is what to do if the type doesn't have this optional additional "hint"?  It requires module designers to think about encodings they otherwise don't really care about.

**Comment by rgwilton** [comment]
We have similar issues where numerical values are being encoded as strings because the owner of those types want them to be displayed in a particular way.  MAC address is one example that I can think of (for which IETF and IEEE now have different string definitions because they have chosen different presentations), Ethertype is another.

One solution to this problem is to be able to define binary encodings of these values.  Another solution is to allow for binary/numeric definitions along with an associated presentation definition.

**Comment by kwatsen** [comment]
The question is if the different types would be selected if binary protocol were known to be used.
For now, we think that issue #51 can be used to add any special annotations.
Closing as not a yang-next issue.

**Comment by kwatsen** [comment]
Per on-list email thread, re-opening and adding to the "Discuss Further" column.

**Comment by vlvassilev** [comment]
Defining new uint* and binary built-in sub-statement {lexical-representation hex-string;} enabling alternative lexical representation might be the right solution:

typedef mac-address {
  type binary {
    length 6;
    lexical-representation hex-string;
  }
}

typedef ether-type {
  type uint16 {
    lexical-representation hex-string;
  }
}

**Comment by abierman** [comment]
YANG to CBOR RFC is not stable
New work in progress
https://www.potaroo.net/ietf/idref/draft-bormann-cbor-yang-standin/
issues raised in this ticket have been resolved

**Comment by kwatsen** [comment]
Since an *encoding* issue, then not in YANG-next

But Jurgen says adding a discriminator would be helpful, then is a YANG-next

Per mentions it could be helpful to have tab-units for JSON-encodings too.
